fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePrimaryKeySegmentIndex.java

```diff
@@ -126,10 +134,17 @@ public DocumentIndexEntry(@Nonnull final Tuple primaryKey, @Nonnull final byte[]
      */
     @Nullable
     public DocumentIndexEntry findDocument(@Nonnull DirectoryReader directoryReader, @Nonnull Tuple primaryKey) {
+        final AtomicReference<DocumentIndexEntry> doc = new AtomicReference<>();
+        directory.getAgilityContext().accept(aContext -> findDocument(aContext, doc, directoryReader, primaryKey));
```

Comment:
Here also, if you pull the `asyncToSync` up to here, you can use `apply` and remove the `AtomicReference`

Reply:
Same here - apply caused cancellation exception.